File: TopCoder/src/main/java/com/leetcode/ValidParenthesis.java
package com.leetcode;

import java.util.Stack;

//https://leetcode.com/problems/valid-parentheses/
public class ValidParenthesis {
    public boolean isValid(String s) {

        Stack<Character> stack = new Stack<>();

        for ( int i = 0 ; i < s.length() ; i++) {
            if (s.charAt(i) == ')') {
                if (stack.isEmpty() || stack.peek() != '(') {
                    return false;
                } else {
                    stack.pop();
                }
            } else if (s.charAt(i) == '}') {
                if (stack.isEmpty() || stack.peek() != '{') {
                    return false;
                } else {
                    stack.pop();
                }
            } else if (s.charAt(i) == ']') {
                if (stack.isEmpty() || stack.peek() != '[') {
                    return false;
                } else {
                    stack.pop();
                }
            } else {
                stack.push(s.charAt(i));
            }
        }
        return stack.isEmpty();
    }

    public static void main(String[] args) {
        String str = "(]";
        boolean result = new ValidParenthesis().isValid(str);
        System.out.println(result);
    }
}

File: TopCoder/python/misc/ValidParenthesis.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        for ss in s:
            if ss is '}':
                if len(stack) > 0 and stack[len(stack) - 1] != '{':
                    return False
                elif len(stack) == 0:
                    return False
                else:
                    stack.pop()
            elif ss is ']':
                if len(stack) > 0 and stack[len(stack) - 1] != '[':
                    return False
                elif len(stack) == 0:
                    return False
                else:
                    stack.pop()
            elif ss is ')':
                if len(stack) > 0 and stack[len(stack) - 1] != '(':
                    return False
                elif len(stack) == 0:
                    return False
                else:
                    stack.pop()
            else:
                stack.append(ss)
        return len(stack) == 0

if __name__ == '__main__':
    sol = Solution()
    res = sol.isValid('[[')
    print(res)


